crates/circuit/src/packed_instruction.rs

```diff
@@ -702,10 +702,91 @@ pub struct PackedInstruction {
     /// feature.
-    pub py_op: OnceLock<Py<PyAny>>,
+    py_op: OnceLock<Py<PyAny>>,
+}
+
+// Custom implementation of `Clone` which removes the ability to clone the underlying `py_op` cache.
+// If a user were to want to clone the cache along with the operation, the user should call
+// [PackedOperation::shallow_clone].
+impl Clone for PackedInstruction {
+    fn clone(&self) -> Self {
+        Self {
+            op: self.op.clone(),
+            clbits: self.clbits,
+            label: self.label.clone(),
+            params: self.params.clone(),
+            qubits: self.qubits,
+            #[cfg(feature = "cache_pygates")]
+            py_op: Default::default(),
+        }
+    }
 }
 
 impl PackedInstruction {
+    /// Creates a new instance of [PackedInstruction] based on an operation.
+    ///
+    /// # Arguments
+    ///
+    /// * `op` - An instance of [PackedOperation] or anything ranging from
+    ///   [StandardGate], [StandardInstruction], [UnitaryGate], [PyGate], [PyInstruction],
+    ///   [PyOperation].
+    /// * `qubits` - The indices of the circuit qubits this instance operates on.
+    /// * `clbits` - The indices of the circuit clbits this instance operates on.
+    /// * `params` - The list of parameters stored by this instruction.
+    /// * `label` - The label of this instruction, if any.
+    pub fn new<O: Into<PackedOperation>>(
+        op: O,
+        qubits: Interned<[Qubit]>,
+        clbits: Interned<[Clbit]>,
+    ) -> Self {
+        Self {
+            op: op.into(),
+            qubits,
+            clbits,
+            params: None,
+            label: None,
+            #[cfg(feature = "cache_pygates")]
+            py_op: OnceLock::new(),
+        }
+    }
+
+    /// Consumes the origal instance and returns another [PackedInstruction] with new parameters.
+    #[inline]
+    pub fn with_params(mut self, params: SmallVec<[Param; 3]>) -> Self {
+        self.params = (!params.is_empty()).then_some(Box::new(params));
+        #[cfg(feature = "cache_pygates")]
+        self.py_op.take();
+        self
+    }
+
+    /// Consumes the origal instance and returns another [PackedInstruction] with a new label.
+    #[inline]
+    pub fn with_label(mut self, label: String) -> Self {
+        self.label = Some(Box::new(label));
+        #[cfg(feature = "cache_pygates")]
+        self.py_op.take();
+        self
+    }
+
+    /// Consumes the current instance of [PackedInstruction] and adds a python
+    /// cached gate.
+    ///
+    /// # Arguments
+    /// * `py_op` - The cached python operation object.
+    #[cfg(feature = "cache_pygates")]
+    pub fn with_py_cache(mut self, py_op: PyObject) -> Self {
+        self.py_op = py_op.into();
+        self
+    }
+
+    pub fn replace_qubits(&mut self, qubits: Interned<[Qubit]>) {
+        self.qubits = qubits;
+    }
+
+    pub fn replace_clbits(&mut self, clbits: Interned<[Clbit]>) {
+        self.clbits = clbits;
+    }
```

Comment:
If there's no logic in the getters/setters of `qubits` and `clbits` and they _have_ getters and setters (which they absolutely should - it makes several things more efficient), then can we just leave them as `pub` fields in the struct?  I'm not sure I see an advantage to the churn of changing things here.